apps/mccurtain/mccurtain.1
.Dd April 16, 2024
.Dt mccurtain 1
.Os mcplex.net
.Sh NAME
.Nm mccurtain
.Nd query for prefixes and origin ASes of IP address, or prefixes of origin AS
.Sh SYNOPSIS
.Nm
.Op Fl d
.Op Fl h Ar mccurtaind_host(s)
.Op Fl p Ar port
.Cm ipv4_address
.Nm
.Op Fl d
.Op Fl v
.Op Fl h Ar mccurtaind_host(s)
.Op Fl p Ar port
.Cm AS_number
.Nm
.Cm -V
.Sh DESCRIPTION
.Nm
is a command-line client that asks mccurtaind(8) for prefixes
covering a given IP address and the corresponding origin AS (autonomous
systems) announcing those prefixes.  It can also ask mccurtaind(8) for
all prefixes originating from a given AS.
.Pp
.Nm
uses libDwmCredence for authentication and encryption.  Hence you
will need to add the mccurtaind(8) public key to your credence known_keys
file, and of course mccurtaind(8) will need your public key in its
known_keys file.  If you don't have a credence keypair, see the
credence(1) manpage.

The following command line options are available:
.Pp
.Bl -tag -width indent
.It Fl h Ar mccurtaind_host(s)
Specify the
.Xr mccurtaind 8
hosts to be queried.  This is a comma-separated list of hostnames and/or
IP addresses,  which will be tried in order.  The first successful
connection will be used.
.It Fl p Ar port
Specify the port that
.Xr mccurtaind 8
is listening on.  The default is port 2126.
.It Fl d
Enable debug level logging on stderr.
.It Fl v
When querying an AS number, print the name and country of the AS and
indent the prefixes by 2 spaces.
.El

.Sh EXAMPLES
The following example requests information for 192.172.226.1.  A single
prefix is returned (192.172.226.0/24), with two origin autonomous systems
(195 and 1909).
.Bd -literal
    % mccurtain 192.172.226.1
    192.172.226.0/24:
      195        US SDSC-AS
      1909       US CAIDA
.Ed

The next example requests information for AS 195.  The prefixes for which
AS 195 is the origin AS are returned.
.Bd -literal
    % mccurtain -v 195
    195        US SDSC-AS
      44.8.5/24
      132.249/16
      137.164.23.42/31
      137.164.23.128/29
      153.105/16
      192.12.207/24
      192.31.21/24
      192.67.81/24
      192.67.82/24
      192.172.226/24
      192.207.234/24
      198.17.46/23
      198.17.101/24
      198.187.222/24
      198.188.136.0/26
      198.202.64/18
      199.165.16/20
.Ed
.Sh ENVIRONMENT
The MCCURTAIND environment variable can be set to the name of the hosts
running
.Xr mccurtaind 8 .
It will be used when the \fI-h mccurtaind_host(s)\fR option is not specified
on the command line.
.Sh FILES
.Bl -tag -width indent
.It Pa ${HOME}/.credence/id_ed25519
The user's credence private key file, created with
.Xr credence 1 .
This file should be owned by the user and have permissions 0600.
It must contain the private part of an Ed25519 key pair.
.It Pa ${HOME}/.credence/id_ed25519.pub
The user's credence public key file, created with
.Xr credence 1 .
This file should be owned by the user and have permissions 0600.
It must contain the public part of an Ed25519 key pair.
.It Pa ${HOME}/.credence/known_keys
The user's credence known keys file.
This file must contain the public keys of services the user will access.
Each public key is used to authenticate a remote service during
authentication.
.It Pa <service>/known_keys
A service utilizing libDwmCredence will have a file containing the public keys
of those allowed to access the service.
The location of this file is service dependent.
.El
.Sh SEE ALSO
.Xr mccurtaind 8 ,
.Xr mkcurtaindb 8 ,
.Xr credence 1
.Sh AUTHORS
Copyright \(co 2024
.An Daniel W. McRobb Aq Mt dwm@mcplex.net


ChangeLog
2024-05-14  dwm  <dwm@nomnom.mcplex.net>

	* tagged as mccurtain-0.0.8

	* etc/asn-list.json: Added AS 136052.

2024-05-14  Daniel W. McRobb  <dwm@www.mcplex.net>

	* tagged as mccurtain-0.0.7

	* etc/asn-list.json: more ASes added.

2024-05-13  Daniel McRobb  <dwm@spark.mcplex.net>

	* tagged as mccurtain-0.0.6

	* apps/mccurtain/mccurtain.cc: Added port number command line option.

2024-05-09  Daniel McRobb  <dwm@toto.mcplex.net>

	* tagged as mccurtain-0.0.5

	* etc/asn-list.json: more ASes added.

2024-05-07  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.4

	* packaging/fbsd_manifest.in: Correct libDwm version to 0.9.23

2024-05-07  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as mccurtain-0.0.3

	* apps/mccurtain/Makefile: include manpage in tarprep target.

	* tagged as mccurtain-0.0.2

	* configure.ac, configure: We need at least version 0.9.23 of libDwm.

	* tagged as mccurtain-0.0.1




apps/mccurtain/mccurtain.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2024
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mccurtain.cc
//!  \author Daniel W. McRobb
//!  \brief NOT YET DOCUMENTED
//---------------------------------------------------------------------------

extern "C" {
  #include <unistd.h>
}

#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <regex>
#include <thread>

#include "DwmSysLogger.hh"
#include "DwmCredencePeer.hh"
#include "DwmMcCurtainRequests.hh"
#include "DwmMcCurtainResponses.hh"
#include "DwmMcCurtainVersion.hh"

using namespace std;
using namespace Dwm;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool GetPeer(const string & host, uint16_t port, Credence::Peer & peer)
{
  bool  rc = false;
  if (peer.Connect(host, port)) {
    Credence::KeyStash   keyStash;
    Credence::KnownKeys  knownKeys;
    if (peer.Authenticate(keyStash, knownKeys)) {
      rc = true;
    }
    else {
      peer.Disconnect();
      Syslog(LOG_ERR, "Failed to authenticate with %s", host.c_str());
    }
  }
  else {
    Syslog(LOG_ERR, "Failed to connect to %s port %hu", host.c_str(), port);
      peer.Disconnect();
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool GetPeer(const vector<string> & hosts, uint16_t port,
                    Credence::Peer & peer)
{
  bool  rc = false;
  for (const auto & host : hosts) {
    if (GetPeer(host, port, peer)) {
      rc = true;
      break;
    }
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void
PrintIpv4AddrResponse(const Dwm::McCurtain::Ipv4AddrResponse & resp)
{
  for (const auto & entry : resp) {
    cout << std::get<0>(entry) << ":\n";
    for (const auto & ase : std::get<1>(entry)) {
      cout << "  " << setiosflags(ios::left) << setw(10) << ase.first << ' '
           << setw(2) << ase.second.CountryCode() << ' '
           << ase.second.Name() << '\n';
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void
PrintASPrefixesResponse(const Dwm::McCurtain::ASPrefixesResponse & resp,
                        bool verbose)
{
  if (verbose) {
    cout << setiosflags(ios::left) << setw(10) << std::get<0>(resp) << ' '
         << setw(2) << std::get<1>(resp).CountryCode() << ' '
         << std::get<1>(resp).Name() << '\n';
  }
  for (const auto & pfx : std::get<2>(resp)) {
    if (verbose) {
      cout << "  ";
    }
    cout << pfx.ToShortString() << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static vector<string> SplitArg(const string & arg)
{
  vector<string>  rc;
  size_t   startIndex = 0;
  size_t   foundIndex;
  while ((foundIndex = arg.find_first_of(',',startIndex)) != string::npos) {
    string  argpiece = arg.substr(startIndex, foundIndex - startIndex);
    if (! argpiece.empty()) {
      rc.push_back(arg.substr(startIndex, foundIndex - startIndex));
    }
    startIndex = foundIndex + 1;
  }
  if (startIndex < arg.size()) {
    string  argpiece = arg.substr(startIndex);
    if (! argpiece.empty()) {
      rc.push_back(arg.substr(startIndex));
    }
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  cerr << "Usage: " << argv0 << " [-d] [-h mccurtaind_host] [-p port] ipv4addr\n"
       << "       " << argv0 << " [-d] [-v] [-h mccurtaind_host] [-p port] AS_number\n"
       << "       " << argv0 << " -V\n\n"
       << "  Note: MCCURTAIND environment variable will be used if\n"
       << "        '-h mccurtaind_host' option is not specified.\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  extern int        optind;
  int               optChar;
  string            hostList;
  uint16_t          port = 2126;
  bool              verbose = false;
  
  Dwm::SysLogger::Open("mccurtain", LOG_PERROR, LOG_USER);
  Dwm::SysLogger::MinimumPriority(LOG_ERR);
  Dwm::SysLogger::ShowFileLocation(true);

  char *mccurtaindEnv = getenv("MCCURTAIND");
  if (nullptr != mccurtaindEnv) {
    hostList = mccurtaindEnv;
  }
  
  while ((optChar = getopt(argc, argv, "dh:p:vV")) != -1) {
    switch (optChar) {
      case 'd':
        Dwm::SysLogger::MinimumPriority(LOG_DEBUG);
        break;
      case 'h':
        hostList = optarg;
        break;
      case 'p':
        try {
          port = stoul(optarg, nullptr, 10);
        }
        catch (...) {
          cerr << "Invalid port " << optarg << '\n';
          Usage(argv[0]);
          return 1;
        }
        break;
      case 'v':
        verbose = true;
        break;
      case 'V':
        cout << Dwm::McCurtain::Version.Version() << '\n';
        return 0;
        break;
      default:
        Usage(argv[0]);
        return 1;
        break;
    }
  }

  if (hostList.empty()) {
    Usage(argv[0]);
    return 1;
  }

  vector<string>  hosts = SplitArg(hostList);
  
  if (optind >= argc) {
    Usage(argv[0]);
    return 1;
  }
  
  Credence::Peer  peer;
  if (GetPeer(hosts, port, peer)) {
    string  arg(argv[optind]);
    if (arg.find_first_of('.') != string::npos) {
      Dwm::McCurtain::Request  req{Dwm::Ipv4Address(arg)};
      if (peer.Send(req)) {
        Dwm::McCurtain::Ipv4AddrResponse  resp;
        if (peer.Receive(resp)) {
          PrintIpv4AddrResponse(resp);
          return 0;
        }
      }
    }
    else {
      try {
        uint32_t  asNum = stoul(arg);
        Dwm::McCurtain::Request  req{asNum};
        if (peer.Send(req)) {
          Dwm::McCurtain::ASPrefixesResponse  resp;
          if (peer.Receive(resp)) {
            PrintASPrefixesResponse(resp, verbose);
            return 0;
          }
        }
      }
      catch (...) {
      }
    }
  }
  return 1;
}

